Src/Web/Web.Shared.Net/RequestTrackingTelemetryModule.cs

```diff
@@ -402,69 +402,100 @@ private class ChildRequestTrackingSuppressionModule
+                    if (!this.IsRequestKnown(rootRequestId))
                     {
-                        if (x.Value < timeout)
-                        {
-                            if (activeRequests.TryRemove(x.Key, out var value))
-                            {
-                                removeCount++;
-                            }
-                        }
+                        // doesn't exist add to A and return true;
+                        this.AddRequestToDictionary(rootRequestId);
+                        return true;
                     }
+                }
+
+                return false;
+            }
+
+            /// <summary>
+            /// Tag new requests.
+            /// Transfer Ids to parent requests.
+            /// </summary>
+            private string TagRequest(HttpContext context)
+            {
+                var newId = Guid.NewGuid().ToString();
+
+                var headers = context.Request.Headers;
 
-                    WebEventSource.Log.ChildRequestTrackingClearingActiveRequests(MAXSIZE, TIMEOUTSECONDS, removeCount);
+                if (headers[HeaderRootRequestId] == null)
+                {
+                    headers[HeaderRootRequestId] = newId;
                 }
 
-                if (context.Request.Headers[RequestIdHeader] == null)
+#if DEBUG
+                // additional ids help developer watch request hierarchy while debugging.
+                if (headers[HeaderRequestId] != null)
                 {
-                    string requestId = Guid.NewGuid().ToString();
-                    context.Request.Headers[RequestIdHeader] = requestId;
-                    activeRequests.TryAdd(requestId, DateTime.UtcNow.Ticks);
+                    headers[HeaderParentRequestId] = headers[HeaderRequestId];
                 }
+
+                headers[HeaderRequestId] = newId;
+#endif
+
+                return newId;
             }
-            
+
             /// <summary>
-            /// Will compare a request id against a hash-set of active requests.
-            /// Will return true if request id is matched, and remove id from hash-set.
-            /// Additional requests with the same id will return false.
+            /// Has this request been tracked.
             /// </summary>
-            internal bool ShouldTrackRequest(HttpContext context)
+            private bool IsRequestKnown(string requestId)
             {
-                string requestId = context.Request.Headers[RequestIdHeader];
+                return activeRequestsA.TryGetValue(requestId, out bool valueA) || activeRequestsB.TryGetValue(requestId, out bool valueB);
+            }
 
-                if (requestId != null && activeRequests.ContainsKey(requestId))
+            /// <summary>
+            /// Track this requestId.
+            /// </summary>
+            private void AddRequestToDictionary(string requestId)
+            {
+                // TODO: LOCK
```

Comment:
Can this be implemented with `Interlocked.CompareExchange()` to avoid lock and still be thread-safe?

Reply:
TL;DR: Nope.

Discussed with Dmitry offline, I'm going to avoid this suggestion.

CompareExchange involves a; target, compareEqualityValue, and newValue.
That would wipe out dictionary A before I could set it to B.
And I can't have the B be a target, based on a comparison of dictionary A.

Also, the strict equality comparison makes me nervous because if two concurrent threads are able to write to dictionary A and push it over the MAX, we would never catch this condition.